Make databricks sql hook return a namedtuple

Joffreybvn opened the pull request:
This is a fix proposal for https://github.com/apache/airflow/pull/36161#pullrequestreview-1779199804

The [ADR for Airflow's DB API](https://github.com/apache/airflow/blob/main/airflow/providers/common/sql/doc/adr/0002-return-common-data-structure-from-dbapihook-derived-hooks.md) specifies that DbApiHooks needs to return a tuple / list of tuples. The databricks hook by defaults returns a dict-like [`databricks.Row`](https://github.com/databricks/databricks-sql-python/blob/62eb1d407ab03fc2c57a6626bec2173dbdd8bbb3/src/databricks/sql/types.py#L24) object which we now transform into a new nametuple - _which is compliant with the 'common data structure' specification._

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

joffreybienvenu-infrabel committed:
feat: Databricks sql hook returns json-serializable namedtuple

NamedTuple can be used as a tuple or a dict. databricks.Row object is a dict-like object. This should break fewer use-cases than switching directly to a tuple/list object.

joffreybienvenu-infrabel committed:
fix: Ignore mypy

namedtuple has to be named "Row", to be similar to the one returned by the Hook. But its variable name should not be Row in this context, to differentiate the real "databricks.Row" object from the namedtuple one.

potiuk reviewed (approved):
Looks good

utkarsharma2 commented on airflow/providers/databricks/hooks/databricks_sql.py:
```diff
@@ -243,11 +244,15 @@ def run(
         if isinstance(result, list):
-            return [list(row) for row in result]
+            columns = result[0].__fields__
+            row_object = namedtuple("Row", columns)
```
The same testcase from earlier PR would still fail here, right? 

```
statement = f"DESCRIBE TABLE {table.name};"
hook.run(statement, parameters=parameters, handler=lambda x: x.fetchall())
```
with this code hook.run() will return 
[SerializableRow(),SerializableRow(), SerializableRow()] instead of [row(), row(), row()]

Since `Row` is not the same as `namedtuple` class this is still a breaking change.

phanikumv replied:
Agree with @utkarsharma2  comment, we still need to update the code in this PR.

potiuk replied:
yep

Joffreybvn replied:
Indeed, this PR is not about preventing the breaking change. But a proposal to break less workflows - _the simple tuple/list is too agressive_ - and reach the goal of [this ADR](https://github.com/apache/airflow/blob/main/airflow/providers/common/sql/doc/adr/0002-return-common-data-structure-from-dbapihook-derived-hooks.md).

The databricks.sql.Row object is a subclass of tuple, which implements:
- a dict-like interface -> breaking change - not available in a namedtuple
- a 'asDict()' method -> breaking change - but namedtuple implements ['_asdict()'](https://docs.python.org/3/library/collections.html#collections.somenamedtuple._asdict)
- fields accessible by name -> supported by namedtuple

Thus users of the dict interface have to rewrite their code. With a major version bump of the provider, would you agree on this change ?

potiuk replied:
Yes. We should avoid breaking change like that

Joffreybvn replied:
I put the serializable namedtuple behind a `return_serializable` flag, which is False by default. With a warning for a future release of the provider, where it would turn to True by default. And eventually later the flag can disappear too to only return the namedtuple.

What do you think ?

phanikumv reviewed (changes requested):
We need to address https://github.com/apache/airflow/pull/36205#discussion_r1425309132

bolkedebruin commented on airflow/providers/databricks/hooks/databricks_sql.py:
```diff
@@ -243,11 +244,15 @@ def run(
     @staticmethod
     def _make_serializable(result):
-        """Transform the databricks Row objects into JSON-serializable lists."""
+        """Transform the databricks Row objects into JSON-serializable namedtuple."""
```
Please remove the JSON remark. It is either XCom serializable, serde serializable or just serializable. JSON has no meaning here.

potiuk replied:
> I wonder btw if we should not have a typedef for it if this is part the the DbAPI. That would make things conceptually stronger and less ambiguous.

Yes. It was missing from the beginning. Agree JSON in the description is wrong.

Joffreybvn replied:
Thanks for the re-scoping ! I tried to be reactive on this one, but did not put enough care in the namings. I updated the code and the PR message.

bolkedebruin reviewed (changes requested):
Given all the discussions around this earlier I'm a bit picky on the naming. We are not making something JSON serializable - but according to the earlier discussion we make the SQL hook return the specified NamedTuple format. 

I wonder btw if we should not have a typedef for it if this is part the the DbAPI. That would make things conceptually stronger and less ambiguous. 

If you insist on naming it something serializable please leave out JSON. It has nothing to do with it.

Can we further be clear on what this PR wants to achieve in the commit message? The title and message are now confusing.

My suggestion:

```
Return SerializableRow in databricks sql hook

The ADR for Airflow' s DB API specifies it needs to return a 
named tuple SerialzableRow or a list of them.  The databricks
hook by defaults returns a dict which we now transform to a
SerializableRow.
```

joffreybienvenu-infrabel committed:
fix: correctly describe serialization

joffreybienvenu-infrabel committed:
feat: disable the return of serializable namedtuple by default

joffreybienvenu-infrabel committed:
fix: make DatabricksSqlOperator return serializable object by default

This operator was already returning a serializable object since #31780. But the PR #31780 was reverted in favor of a serialization at the Hook level. To avoid regression, the DatabricksSqlOperator has to return a serializable

joffreybienvenu-infrabel committed:
fix: add `return_serializable` to databricks operator tests

joffreybienvenu-infrabel committed:
fix: docstring spellchecks

potiuk reviewed (approved):
I like the way it is now.

bolkedebruin commented on airflow/providers/databricks/hooks/databricks_sql.py:
```diff
@@ -80,8 +85,18 @@ def __init__(
+        self.return_serializable = return_serializable
         self.additional_params = kwargs
 
+        if not self.return_serializable:
```
Who will set this and why and when? Is this a kind of backwards compatibility? Will this not generate a lot of warnings?

Joffreybvn replied:
Yes, this is to ensure the change is backward compatible.

Users of the hook will be warned. Users of the `DatabricksSqlOperator` won't, because the operator was already returning a list of tuple - which becomes a list of namedtuple with this PR. Thus, it will generate noise only to hook users.

The goal is to effectively do a breaking change later. But if we don't want that - or not _that_ soon, I can check to print the warning only when the provider loads ?

Joffreybvn replied:
Will clarify the docstring and the warning

joffreybienvenu-infrabel committed:
fix: rewrite docstring and warning message

joffreybienvenu-infrabel committed:
fix: docstring spellchecks

bolkedebruin commented on airflow/providers/databricks/hooks/databricks_sql.py:
```diff
@@ -241,13 +258,17 @@ def run(
-            return [list(row) for row in result]
-        elif isinstance(result, Row):
-            return list(result)
+    def _make_serializable(self, result):
```
Can we have this type hinted? Why is it not a static method anymore?

Joffreybvn replied:
Not a staticmethod anymore because checking self.return_serializable is needed. I added typings

bolkedebruin commented on airflow/providers/databricks/hooks/databricks_sql.py:
```diff
@@ -241,13 +258,17 @@ def run(
+            elif isinstance(result, Row):
+                columns = result.__fields__
+                return namedtuple("Row", columns)(*result)
         return result
```
If something else else a list / row gets in this implicitly returns what was given. Do we want that?

Joffreybvn replied:
Not a strong opinion, but I think we do want that:

The `result` input of this function is the output of the [`handler()`](https://github.com/apache/airflow/blob/280b8b30061af2dec60ac83f9fafb92b3f0663c1/airflow/providers/databricks/hooks/databricks_sql.py#L243C58-L243C65) method. This method can be one of the default ones of Airflow (doing simply *fetchone()* or *fetchall()*), but also a totally custom handler injected via the `handler` parameter. Maybe the user want to do something else/something more after fetching, which returns something else than a Row / list[Row] ? Even if it's not the right place to transform the data IMO, blocking anything else than Row/list[Row] can potentially break workflows.

I will improve the isinstance on the list to check if this list contains Row object. And otherwise, the list get ignored too and returned as-is.

Joffreybvn replied:
The same is applied to the ODBCHook

joffreybienvenu-infrabel committed:
feat: rename `make_serializable` into `make_common_data_structure` for clarity

Apply the ADR 0002 - https://github.com/apache/airflow/blob/main/airflow/providers/common/sql/doc/adr/0002-return-common-data-structure-from-dbapihook-derived-hooks.md

joffreybienvenu-infrabel committed:
feat: Add typing for `_make_common_data_structure` in databricks

joffreybienvenu-infrabel committed:
feat: Apply typing to ODBC hook

joffreybienvenu-infrabel committed:
fix: patch pyodbc.Row to make isinstance() checks pass with row_mock object

joffreybienvenu-infrabel committed:
fix: Use List in type casting for python38 compatibility

joffreybienvenu-infrabel committed:
feat: Remove all mentions of serialization in databricks hook

joffreybienvenu-infrabel committed:
feat: Rename `_make_serializable` into `_make_common_data_structure` in static checks

joffreybienvenu-infrabel committed:
fix: use `return_tuple` in databricks tests

joffreybienvenu-infrabel committed:
fix: add deprecated _make_serializable method in odbc.py

joffreybienvenu-infrabel committed:
bump: min sql.common to 1.9.1 for odbc and databricks

potiuk commented on airflow/providers/databricks/provider.yaml:
```diff
@@ -59,7 +59,7 @@ versions:
 dependencies:
   - apache-airflow>=2.6.0
-  - apache-airflow-providers-common-sql>=1.8.1
+  - apache-airflow-providers-common-sql>=1.9.1
```
```suggestion
  - apache-airflow-providers-common-sql>=1.9.0
```

UPDATE: nope, it's good.

potiuk replied:
Ah no.... scratch that... 1.9.1 is fine (looked at all version of provider.yaml of common.sql). Can you, however add 1.9.1 in the `provider.yaml` as it will not pass the tests. When two providers depends on each other with min-version, the min-version has to be added to provider.yaml of the provider that is depended on.

potiuk commented on airflow/providers/odbc/hooks/odbc.py:
```diff
@@ -212,17 +216,27 @@ def get_sqlalchemy_connection(
-    @staticmethod
-    def _make_serializable(result: list[pyodbc.Row] | pyodbc.Row | None) -> list[NamedTuple] | None:
-        """Transform the pyodbc.Row objects returned from an SQL command into JSON-serializable NamedTuple."""
+    def _make_serializable(self, result: Any) -> Any:
```
Actually what we would have to do, to not yank the current odbc provider is to add code in the common.sql to check if it's there and call it. Adding it here while we also add >= 1.9.1 on the common.sql makes no sense (because 1.9.1 will never call it)

The back-compatibilty here is to let the "just released" odbc provider to work with the new `"common.sql" - not the other way round.

But maybe better - just remove it and I will simply yank the odbc provider we just released, it's going to be way simpler @Joffreybvn

Joffreybvn replied:
I think we need this back-compat fix anyway because the `make_serializable´ was introduced in odbc two patches ago (in [4.2.0](https://airflow.apache.org/docs/apache-airflow-providers-odbc/4.3.0/changelog.html#id2)).

Thus, I added it with a warning. Airflow 2.8.1 constraint will need this latest odbc provider version to not annoy users with the warning. So that only people who stick with odbc==4.2.0 or odbc==4.3.0 while upgrading common.sql get warned.

joffreybienvenu-infrabel committed:
fix: add version 1.9.1 to provider.yaml of common.sql

joffreybienvenu-infrabel committed:
fix: move back-compat in common.sql

potiuk reviewed (approved):
Very cool! Thanks. 

Yeah. Providers' coupling vs. back-compat makes things quite a bit more difficult 

Thanks for being responsive and keep up with maintainers having different opinions and arguments

bolkedebruin commented on airflow/providers/databricks/hooks/databricks_sql.py:
```diff
@@ -241,13 +267,23 @@ def run(
+        # Below ignored lines respect namedtuple docstring, but mypy do not support dynamically
+        # instantiated namedtuple, and will never do: https://github.com/python/mypy/issues/848
+        if self.return_tuple:
+            if isinstance(result, list) and all(isinstance(item, Row) for item in result):
```
Potentially you are checking thousands of rows here and `isinstance` is quite slow as it needs to walk the inheritance tree. Is there a way to check a attribute(s) to verify you are dealing with a Row?

potiuk replied:
What would be your suggestion here @bolkedebruin ? I guess it is not blocking but simple optimisation suggestion? If so - is that fine if we separate it as a follow up issue to optimize it? I guess this is not something that BLOCKS this pr? 

I think we can definitely optimise it in the future, however, again - perfect is the enemy of good and it's holding back release of a new waves of providers . My suggestion is to get it out now and separate it as a follow-up PR.

Is that OK @bolkedebruin ?

bolkedebruin replied:
an `isinstance` check can take between 15%-60% more time than an attribute check. This change affects current operations, so maybe this is an pre-mature optimization or maybe we do want to apply it. I leave this to @Joffreybvn as he is using it I assume.

bolkedebruin commented on airflow/providers/databricks/hooks/databricks_sql.py:
```diff
@@ -241,13 +267,23 @@ def run(
-            return list(result)
+    def _make_common_data_structure(
+        self, result: list[T] | T | None
+    ) -> list[T] | T | list[tuple] | tuple | None:
```
As explained in the comment `list[T] | T` does not specify a common format. It just takes the input and returns the same. That does not make sense. `list[tuple] | tuple` should be sufficient. And I'd prefer it to be without `None` as that seems not compatible with a 'common_data_structure`

potiuk replied:
As explained already in previous discussions I plan to add it in a follow up PR - together with ADR detailing some more of it. I think we should take a few more steps to clarify it and I even have issue defined here and assigned to me: https://github.com/apache/airflow/issues/36224

Perfect is an enemy of good. I suggest we complete this thing now and then add typing as seperate PR.

Is that ok @bolkedebruin  ? Can we do it this way and not circle back the same comment in this PR?

Joffreybvn replied:
I refactored `_make_common_data_structure` with the suggested typing, without None.

joffreybienvenu-infrabel committed:
feat: Implement strict `tuple`/`list[tuple]` return in ODBC and Databricks hooks

Joffreybvn commented on airflow/providers/databricks/hooks/databricks_sql.py:
```diff
@@ -167,7 +195,7 @@ def run(
         split_statements: bool = ...,
         return_last: bool = ...,
-    ) -> T | list[T]:
+    ) -> tuple | list[tuple] | list[list[tuple] | tuple] | None:
```
Mypy was not happy with `-> tuple | list[tuple] | list[list[tuple] | tuple]` as output - which is correct. I got error: "*Overloaded function signatures 1 and 2 overlap with incompatible return types*".

[Following mypy docs](https://mypy.readthedocs.io/en/stable/more_types.html#type-checking-the-variants), I added a `None` for it to be stop complaining. But that's not correct, this method cannot return None when a handler is provided...

Implementing definitive typing is out of the scope of this PR, and I admit mypy beat me on that case. Thus I let that for #36224 - but I'm whiling to help on that issue too of course.

bolkedebruin reviewed (approved):
Much much better! I understand what this is about now. The follow up would make things even clearer so absolutely appreciated if you want to help out there.
